man/dump_tape.1.in
.TH dump_tape "1" "@MONTH@ @YEAR@" "mhvtl @VERSION@" "User Commands"
.SH NAME
dump_tape \- Utility to dump tape contents for virtual media in VTL.
.SH SYNOPSIS
.B dump_tape
.B [ \-h ]
.br
.B dump_tape
.B [ \fIOPTIONS\fR]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Dump the contents of the specified PCL. If a library is specified, look
there, else look through all of the libraries.
.SH OPTIONS
.TP
.B \-h
display usage information and exit
.TP
.B \-v
be verbose
.TP
.B \-d
print debugging information
.TP
.BR \-l lib_no
\fB\-l n\fR
Where lib_no is the library index number (default library index numbers 10 & 30).
.TP
\fB\-m PCL\fR
where PCL is the Physical Cartridge Label (barcode). This is required.
.TP
.B \-D
dump data
.SH AUTHOR
Written by Mark Harvey
.SH BUGS
Needs to be made user friendly.
.SH "REPORTING BUGS"
Report bugs to <markh794@gmail.com>
.SH COPYRIGHT
Copyright \(co 2005 Free Software Foundation, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.BR make_vtl_media(1),
.BR library_contents(5),
.BR vtlcmd(1),
.BR vtllibrary(1),
.BR vtltape(1),
.BR edit_tape(1),
.BR preload_tape(1)


ChangeLog
* Thu Oct 07 10:40:00 AEST 2021 Mark Harvey <markh794@gmail.com> <mark.harvey@nutanix.com>
- rename list.h to avoid generic name conflict
- rename scsi.h to avoid generic name conflict
- rename log.h to avoid generic name conflict
- Exclude __builtin_cpu_supports() on non x86_64 CPU types
- Fix spelling of retrieving
- mhvtl.ko: Define SG_SEGMENT_SZ only if not defined
- Increase default kmem_cache bounce buffer size to 64k
- mhvtl.ko: kernel module oops on PPC
- mhvtl.ko: Log errors if they occur
- mhvtl.ko: Correct compiler warning about printf var sizes
- mhvtl.ko: Simplify file_inode()
- mhvtl.ko: Use _safe version of list_for_each_entry
- mhvtl.ko: Limit number of outstanding queued commands
- mhvtl.ko: Log minor number with other ioctl info
- mhvtl.ko: clean up noisy logging at level 2
- mhvtl.ko: Initialise outstanding op struct before adding to list
- mktape: Set default version string based from MHVTL_VERSION
- Update LTO-8 media denisty codes
- Security Protocol OUT: Fix null pointer check
- Fix routine to extract barcode from string
- Fix typos in mktape man page
- dump_tape: Fix segfault due to local variable conflict
- Use bounce buffer if tape block size is larger than request buffer
- Remove unused scsi_host_template field
- mhvtl.ko: Compile on RedHat 8.x kernel
- mhvtl.ko: reinstate HAVE_UNLOCKED_IOCTL
- mhvtl.ko: Remove reference to DRIVER_SENSE
- Update function names with 'mhvtl_' prefix
- Update kernel /sys/ location to suit new pseudo name space
- OOM: /proc/<pid>/oom_adj is deprecated. Using oom_score_adj.
- systemd: Update Makefile to include systemd install path
- mhvtl.spec: Update to build a package on CentOS8

* Tue Mar 10 09:00:00 AEST 2020 Mark Harvey <markh794@gmail.com> <mark.harvey@nutanix.com>
- Fix kernel module oops/panic due to race between placing OP ready for user space before all variable were initialised
- Add ccan version of crc32c routine which includes CPU optimisations for sse4.2 capabile CPUs

* Sun Oct 06 12:20:00 AEST 2019 Mark Harvey <markh794@gmail.com> <mark.harvey@nutanix.com>
- Fix kernel copy_to_user()/copy_from_user() to use white-listed buffer
- Fix kernel compile for 5.0+ kernels

* Wed Apr 10 10:30:30 AEST 2019 Mark Marvey <markh794@gmail.com> <mark.harvey@nutanix.com>
- Fix long standing bug on media unload
- Added CRC32 checks for each block written/read
- allow kernel module to compile on 5.0 kernels
- Add example on how to configure LIO pscsi (passthru) driver with mhVTL
- Fix parsing args in dump_tape

* Thu Mar 28 14:30:30 AEST 2019 Mark Marvey <markh794@gmail.com> <mark.harvey@nutanix.com>
- Create virtual media on rpm install
- Fixed a couple typos in spec file (one from several years ago)
- Don’t overwrite /etc/mhvtl/* but install as ‘rpmnew’
- Replace SUSE specific rpm macros with hand crafted ’systemctl’ commands.
- systems-generators dir needed munging for CentOS7
- Include ‘ChangeLog’ as part of the tar ball - the time/date build on man pages is dependant

* Thu Nov 29 15:42:24 PST 2018 Lee Duncan <lduncan@suse.com>
- Many changes to support systemd:
  - Removed /etc/init.d/mhvtl script
  - Config files are now generated at install time, not the first
    time the service is run
  - Created service files and templates, and a generator
  - Updated INSTALL file with sytemd info
  - removed build_library_config and mhvtl init script
  - Updated man pages and doc/index.html
  - Updated make files
  - created generate_device_conf and generate_library_contents, called at
    'make install' time, to take the place of what the init file used to do
  - updated SPEC file for systemd
  - updated example scripts
  - fixed asprintf calls to check return values (to make compiler happy)
  - added some arguments and better arg parsing to vtltape and vtllibray
- moved some .gitignore entries around, to be more local
- Bumped version to 1.5-5
- Fixed config file error in library_contents.* where VERSION was missing

* Thu Sep 7 2014 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.5-2
- Fix vtllibrary segfault using 'default' emulation
- Correct slot mappings for L180/L700 emulation
- Correct slot mappings for SL500 emulation
- Increase max serial number length to 15 chars
- More accurate emulation of IBM TS3100/3200 inquiry page bits
- Fix bug with 'edit_tape' not understanding the 'null' media type

* Thu May 1 2014 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.5-1
- Fix bug with vpd pages 0x80 & 0x83 in library personality modules.

* Sun Apr 13 2014 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.5-0
- Introduce library personality modules - limited emulations
- Resolve bug with more than 1260 slots
- Rename mhvtl bus type so scsi_debug & mhvtl can load together.

* Sun Oct 20 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-10
- Various cleanups due to static code check 'smatch'
- Fix TapeAlert bit offsets
- Return bit/byte offset for sense INVALID FIELD IN CDB/PARAMS
- Fix returned data size when 0 data is requested (kernel module fix)
- SSC - Fix op code 0Fh (read reverse) incorrectly set to 'reserve'
- SMC - Ability to keep library media change persistant across restarts
- SSC - Implement 'read media serial number' op code
- Fix 'make_vtl_media' script to understand IBM 03592 media type

* Thu Aug 29 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-9
- SSC -> MODE SELECT overhaul.
- Sanitize delay values introduced in 1.4-8
- daemons now open /dev/mhvtlXX nodes in exclusive mode
- Code cleanup based on Valgrind - memory leaks and unused file handles
- Correctly set SCSI revision field for T10000B
- SMC -> Adjust behaviour if 'Scalar' library defined
 set 'BarC' field in inquiry page and
 pad "Full Firmware Revision" field with compile time/date

* Sat Jun 29 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-8

usr/dump_tape.c
/*
 *	Dump headers of 'tape' datafile
 *
 * Copyright (C) 2005 - 2009 Mark Harvey markh794 at gmail dot com
 *                                mark.harvey at nutanix dot com
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#define _FILE_OFFSET_BITS 64

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <inttypes.h>
#include <zlib.h>
#include "minilzo.h"
#include "be_byteshift.h"
#include "mhvtl_scsi.h"
#include "mhvtl_list.h"
#include "vtl_common.h"
#include "vtllib.h"
#include "vtltape.h"
#include "q.h"
#include "ssc.h"
#include "ccan/crc32c/crc32c.h"

char mhvtl_driver_name[] = "dump_tape";
int verbose = 0;
int debug = 0;
long my_id = 0;
int lib_id;
struct priv_lu_ssc lu_ssc;
struct lu_phy_attr lunit;

extern char home_directory[HOME_DIR_PATH_SZ + 1];

static char *progname;

static void print_mam_info(void)
{
	uint64_t size;
	uint64_t remaining;
	char size_mul;		/* K/M/G/T/P multiplier */
	char remain_mul;	/* K/M/G/T/P multiplier */
	int a;
	static const char mul[] = " KMGT";

	size = get_unaligned_be64(&mam.max_capacity);
	remaining = get_unaligned_be64(&mam.remaining_capacity);

	size_mul = remain_mul = ' ';
	for (a = 0; a < 4; a++) {
		if (size > 5121) {
			size >>= 10;	/* divide by 1024 */
			size_mul = mul[a+1];
		}
	}
	for (a = 0; a < 4; a++) {
		if (remaining > 5121) {
			remaining >>= 10;	/* divide by 1024 */
			remain_mul = mul[a+1];
		}
	}

	printf("Media density code: 0x%02x\n", mam.MediumDensityCode);
	printf("Media type code   : 0x%02x\n", mam.MediaType);
	printf("Media description : %s\n", mam.media_info.description);
	printf("Tape Capacity     : %" PRId64 " (%" PRId64 " %cBytes)\n",
					get_unaligned_be64(&mam.max_capacity),
					size, size_mul);
	switch (mam.MediumType) {
	case MEDIA_TYPE_CLEAN:
		printf("Media type        : Cleaning media\n");
		break;
	case MEDIA_TYPE_DATA:
		printf("Media type        : Normal data\n");
		break;
	case MEDIA_TYPE_NULL:
		printf("Media type        : NULL - Use with caution !!\n");
		break;
	case MEDIA_TYPE_WORM:
		printf("Media type        : Write Once Read Many (WORM)\n");
		break;
	default:
		printf("Media type        : Unknown - please report !!\n");
		break;
	}

	printf("Media             : %s\n",
				(mam.Flags & MAM_FLAGS_MEDIA_WRITE_PROTECT) ?
					"Write-protected" : "read-write");
	printf("Remaining Tape Capacity : %" PRId64 " (%" PRId64 " %cBytes)\n",
				get_unaligned_be64(&mam.remaining_capacity),
				remaining, remain_mul);
}

static int read_data(uint8_t *sam_stat)
{
	uint8_t *p;
	uint32_t ret;

	printf("c_pos->blk_size: %d\n", c_pos->blk_size);

	if (c_pos->blk_size <= 0) {
		printf("Data size: %d - skipping read\n", c_pos->blk_size);
		return 0;
	}
	p = malloc(c_pos->blk_size);
	if (!p) {
		fprintf(stderr, "Unable to allocate %d bytes\n", c_pos->blk_size);
		return -ENOMEM;
	}
	ret = readBlock(p, c_pos->blk_size, 1, sam_stat);
	if (ret != c_pos->blk_size) {
		printf("Requested %d bytes, received %d\n",
				c_pos->blk_size, ret);
	}
	free(p);
	puts("\n");
	return ret;
}

void find_media_home_directory(char *config_directory, char *home_directory, long lib_id);

static void usage(char *errmsg)
{
	if (errmsg)
		printf("%s\n", errmsg);
	printf("Usage: %s OPTIONS\n", progname);
	printf("Where OPTIONS are from:\n");
	printf("  -h               Print this message and exit\n");
	printf("  -d               Enable debugging\n");
	printf("  -v               Be verbose\n");
	printf("  -D               Dump data\n");
	printf("  -l lib_no        Look in specified library\n");
	printf("  -f pcl           Look for specified PCL\n");
	exit(errmsg ? 1 : 0);
}

int main(int argc, char *argv[])
{
	uint8_t sam_stat;
	char *pcl = NULL;
	int rc;
	int libno = 0;
	int indx;
	int dump_data = FALSE;
	char *config = MHVTL_CONFIG_PATH"/device.conf";
	FILE *conf;
	char *b;	/* Read from file into this buffer */
	char *s;	/* Somewhere for sscanf to store results */

	progname = argv[0];

	if (argc < 2)
		usage("Not enough arguments");

#ifdef __x86_64__
	if (__builtin_cpu_supports("sse4.2")) {
		printf("crc32c using Intel sse4.2 hardware optimization\n");
	} else {
		printf("crc32c not using Intel sse4.2 optimization\n");
	}
#endif

	while (argc > 1) {
		if (argv[1][0] == '-') {
			switch (argv[1][1]) {
			case 'h':
				usage(NULL);
				break;
			case 'd':
				debug++;
				verbose = 9;	/* If debug, make verbose... */
				break;
			case 'f':
				if (argc > 1)
					pcl = argv[2];
				else
					usage("More args needed for -f");
				break;
			case 'l':
				if (argc > 1)
					libno = atoi(argv[1]);
				else
					usage("More args needed for -l");
				break;
			case 'D':
				dump_data = TRUE;
				break;
			case 'v':
				verbose++;
				break;
			default:
				usage("Unknown option");
				break;
			}
		}
		argv++;
		argc--;
	}

	if (pcl == NULL)
		usage("No PCL number supplied");

	conf = fopen(config , "r");
	if (!conf) {
		fprintf(stderr, "Cannot open config file %s: %s\n", config,
					strerror(errno));
		exit(1);
	}
	s = malloc(MALLOC_SZ);
	if (!s) {
		perror("Could not allocate memory");
		exit(1);
	}
	b = malloc(MALLOC_SZ);
	if (!b) {
		perror("Could not allocate memory");
		exit(1);
	}

	rc = ENOENT;

	if (libno) {
		printf("Looking for PCL: %s in library %d\n", pcl, libno);
		find_media_home_directory(NULL, home_directory, libno);
		rc = load_tape(pcl, &sam_stat);
	} else { /* Walk thru all defined libraries looking for media */
		while (readline(b, MALLOC_SZ, conf) != NULL) {
			if (b[0] == '#')	/* Ignore comments */
				continue;
			/* If found a library: Attempt to load media
			 * Break out of loop if found. Otherwise try next lib.
			 */
			if (sscanf(b, "Library: %d CHANNEL:", &indx)) {
				find_media_home_directory(NULL, home_directory, indx);
				rc = load_tape(pcl, &sam_stat);
				if (!rc)
					break;
			}
		}
	}

	fclose(conf);
	free(s);
	free(b);

	if (rc) {
		fprintf(stderr, "PCL %s cannot be dumped, "
				"load_tape() returned %d\n",
					pcl, rc);
		exit(1);
	}

	if (lzo_init() != LZO_E_OK) {
		fprintf(stderr, "internal error - lzo_init() failed !!!\n");
		fprintf(stderr,
			"(this usually indicates a compiler bug - try recompiling\nwithout optimizations, and enable '-DLZO_DEBUG' for diagnostics)\n");
		exit(3);
	}
	print_mam_info();

	print_filemark_count();
	if (verbose) {
		printf("Dumping filemark meta info\n");
		print_metadata();
	}

	while (c_pos->blk_type != B_EOD) {
		print_raw_header();
		if (dump_data) {
			read_data(&sam_stat);
		}
		position_blocks_forw(1, &sam_stat);
	}
	print_raw_header();
	unload_tape(&sam_stat);

	return 0;
}
